File: services/web/src/database/migrate.ts
/**
 * Migrador de esquema minimalista (paridad con `services/scraper/src/scraper/migrate.py`).
 *
 * Aplica en orden los ficheros `NNNN_*.sql` de `db/migrations` (SQL neutro, contrato
 * compartido) y registra los aplicados en la MISMA tabla `schema_migrations`, por nombre de
 * fichero. Idempotente y compatible con el migrador Python: cualquiera de los dos servicios
 * puede arrancar el esquema sin pisar al otro.
 *
 * Serializado con un advisory lock que comparte con el migrador Python (#298): los dos leen el
 * conjunto de aplicadas ANTES del bucle, así que sin él dos procesos que arranquen a la vez ven la
 * misma lista de pendientes e intentan aplicar el mismo fichero.
 *
 * Uso: `node dist/database/migrate.js` (o `pnpm migrate`). Config por entorno:
 *   DATABASE_URL             (requerido)
 *   WEB_MIGRATIONS_DIR       (opcional; por defecto db/migrations relativo al repo)
 *   WEB_MIGRATION_LOCK_WAIT  (opcional; segundos de espera del lock, por defecto 300)
 */
import { readdirSync, readFileSync } from 'node:fs';
import { join, resolve } from 'node:path';

import postgres from 'postgres';

const TRACKING_TABLE = `
  CREATE TABLE IF NOT EXISTS schema_migrations (
    version    TEXT PRIMARY KEY,
    applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
  )
`;

/**
 * Identificador del advisory lock que serializa las migraciones. Tiene que ser el MISMO número que
 * `LOCK_MIGRACIONES` en `services/scraper/src/scraper/migrate.py` o los dos migradores no se ven;
 * lo ata el test de paridad de `services/scraper/tests/test_migrate.py`. El valor es `0x64746d67`,
 * que en ASCII es `dtmg` (deal-tracker migrations): arbitrario, pero reconocible en un `pg_locks`.
 *
 * Los advisory locks son POR BASE DE DATOS, no por cluster (verificado el 13/08/2026), así que
 * `deal_tracker`, `deal_tracker_qa` y `deal_tracker_prod` no se estorban entre sí pese a compartir
 * instancia CNPG con otros cuatro proyectos.
 */
export const LOCK_MIGRACIONES = 1685351783;

/**
 * Segundos que se espera el lock. Generoso a propósito: quien lo retiene está aplicando
 * migraciones, y algunas obligan a un `REINDEX` (0014, 0029) que sobre el catálogo entero puede
 * tardar. Lo que esto acota no es contención, es que el otro migrador haya muerto dejando el lock
 * colgado — y ahí 5 minutos ya es un tope. `0` espera lo que haga falta.
 */
const DEFAULT_LOCK_WAIT_SEC = 300;

function migrationsDir(): string {
  const override = process.env.WEB_MIGRATIONS_DIR;
  if (override && override.trim() !== '') {
    return resolve(override.trim());
  }
  // dist/database/migrate.js -> 4 niveles arriba está la raíz del repo (services/web/dist/database).
  return resolve(__dirname, '../../../../db/migrations');
}

function lockWaitSeconds(): number {
  const raw = process.env.WEB_MIGRATION_LOCK_WAIT;
  if (!raw || raw.trim() === '') return DEFAULT_LOCK_WAIT_SEC;
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : DEFAULT_LOCK_WAIT_SEC;
}

/**
 * Toma el lock de migraciones sobre una conexión FIJA, esperando como mucho `waitSec` segundos.
 *
 * El `lock_timeout` va por `set_config(..., is_local => true)` —o sea `SET LOCAL`— dentro de una
 * transacción que se cierra aquí mismo, así que revierte solo y no queda pisado para el resto de
 * la sesión. El lock es de SESIÓN y sobrevive a ese `commit` (comprobado).
 *
 * Ojo con la transacción a mano: `sql.reserve()` devuelve un objeto que en tiempo de EJECUCIÓN no
 * tiene `.begin()` —solo lo lleva el pool—, aunque los tipos de `postgres` digan que
 * `ReservedSql extends Sql`. Verificado con postgres 3.4.9: `typeof reserved.begin === 'undefined'`.
 */
async function tomarLock(reserved: postgres.ReservedSql, waitSec: number): Promise<void> {
  await reserved.unsafe('begin');
  try {
    if (waitSec > 0) {
      await reserved`SELECT set_config('lock_timeout', ${`${Math.round(waitSec * 1000)}ms`}, true)`;
    }
    await reserved`SELECT pg_advisory_lock(${LOCK_MIGRACIONES})`;
    await reserved.unsafe('commit');
  } catch (err) {
    await reserved.unsafe('rollback');
    throw err;
  }
}

export async function runMigrations(sql: postgres.Sql): Promise<string[]> {
  const dir = migrationsDir();
  const files = readdirSync(dir)
    .filter((f) => f.endsWith('.sql'))
    .sort();

  // Una conexión fija para todo: un advisory lock de sesión vive en SU backend, así que si el
  // cuerpo saliera por otra conexión del pool el lock no cubriría nada. Hoy `max: 1` lo hacía
  // cierto por accidente; esto lo hace cierto por construcción.
  const reserved = await sql.reserve();
  try {
    await tomarLock(reserved, lockWaitSeconds());
    try {
      await reserved.unsafe(TRACKING_TABLE);
      const appliedRows = await reserved<{ version: string }[]>`SELECT version FROM schema_migrations`;
      const applied = new Set(appliedRows.map((r) => r.version));

      const justApplied: string[] = [];
      for (const file of files) {
        if (applied.has(file)) continue;
        const text = readFileSync(join(dir, file), 'utf8');
        await reserved.unsafe('begin');
        try {
          await reserved.unsafe(text);
          await reserved`INSERT INTO schema_migrations (version) VALUES (${file})`;
          await reserved.unsafe('commit');
        } catch (err) {
          await reserved.unsafe('rollback');
          throw err;
        }
        justApplied.push(file);
      }
      return justApplied;
    } finally {
      await reserved`SELECT pg_advisory_unlock(${LOCK_MIGRACIONES})`;
    }
  } finally {
    reserved.release();
  }
}

async function main(): Promise<void> {
  const url = process.env.DATABASE_URL;
  if (!url) {
    throw new Error('Falta DATABASE_URL');
  }
  // onnotice compacto: los NOTICE de Postgres (p.ej. "IF NOT EXISTS ... skipping") no deben
  // volcar el objeto entero por consola.
  const sql = postgres(url, {
    max: 1,
    onnotice: (n) => {
      if (n.severity !== 'NOTICE') console.warn(`[pg ${n.severity}] ${n.message}`);
    },
  });
  try {
    const applied = await runMigrations(sql);
    if (applied.length === 0) {
      console.log('Sin migraciones pendientes.');
    } else {
      console.log(`Aplicadas ${applied.length} migración(es): ${applied.join(', ')}`);
    }
  } finally {
    await sql.end();
  }
}

// Solo ejecuta si se invoca directamente (no al importarlo desde un test).
if (require.main === module) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: services/scraper/src/scraper/migrate.py
"""Migrador de esquema minimalista y sin dependencias externas.

Aplica en orden los ficheros `NNNN_*.sql` de `db/migrations` (SQL neutro,
contrato compartido con el servicio web) y registra los aplicados en
`schema_migrations`. Idempotente: reejecutarlo no reaplica nada.

Serializado con un advisory lock que comparte con el migrador del web (#298): los dos leen el
conjunto de aplicadas ANTES del bucle, así que sin él dos procesos que arranquen a la vez ven la
misma lista de pendientes e intentan aplicar el mismo fichero.
"""

from __future__ import annotations

import os
from pathlib import Path

import psycopg
from psycopg.pq import TransactionStatus

# Identificador del advisory lock que serializa las migraciones. Tiene que ser el MISMO número en
# `services/web/src/database/migrate.ts` o los dos migradores no se ven; lo ata el test de paridad
# de `tests/test_migrate.py`. El valor es `0x64746d67`, que en ASCII es `dtmg` (deal-tracker
# migrations): arbitrario, pero reconocible en un `pg_locks` y estable.
#
# Los advisory locks son POR BASE DE DATOS, no por cluster (verificado el 13/08/2026: la misma
# clave se toma sin esperar desde otra base de la misma instancia, y `pg_locks.database` trae el
# OID). Importa porque `deal_tracker`, `deal_tracker_qa` y `deal_tracker_prod` comparten la misma
# CNPG con cuatro proyectos ajenos: esto no los alcanza, y QA no bloquea a prod.
LOCK_MIGRACIONES = 1685351783

# Segundos que se espera el lock. Generoso a propósito y distinto de `SCRAPER_LOCK_TIMEOUT` (30 s,
# #169): quien lo retiene está aplicando migraciones, y algunas obligan a un `REINDEX` (0014, 0029)
# que sobre un catálogo entero no cabe en 30 s. Lo que esto acota no es contención, es que el otro
# migrador haya muerto dejando el lock colgado — y para eso 5 minutos ya es un tope.
DEFAULT_LOCK_WAIT = 300.0


def _default_migrations_dir() -> Path:
    """Ubicación de db/migrations. Configurable por entorno (útil en el contenedor)."""
    override = os.environ.get("SCRAPER_MIGRATIONS_DIR")
    if override:
        return Path(override)
    # services/scraper/src/scraper/migrate.py -> parents[4] == raíz del repo
    return Path(__file__).resolve().parents[4] / "db" / "migrations"


DEFAULT_MIGRATIONS_DIR = _default_migrations_dir()

_TRACKING_TABLE = """
CREATE TABLE IF NOT EXISTS schema_migrations (
    version    TEXT PRIMARY KEY,
    applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
)
"""


def _tomar_lock(conn: psycopg.Connection, espera: float) -> None:
    """Toma el lock de migraciones, esperando como mucho `espera` segundos.

    El `lock_timeout` va con `set_config(..., is_local=true)` —o sea `SET LOCAL`— dentro de una
    transacción que se cierra aquí mismo. Así el valor revierte solo en el `commit` y es imposible
    dejar pisado el `SCRAPER_LOCK_TIMEOUT` que `db.connect()` puso para el resto de la pasada, que
    es la garantía de #169. El lock es de SESIÓN, así que sobrevive a ese `commit` (comprobado).

    Si la espera se agota, Postgres cancela con `55P03` y psycopg lo levanta como
    `LockNotAvailable`, que es justo lo que `run.py` ya captura.
    """
    with conn.cursor() as cur:
        if espera > 0:
            cur.execute("SELECT set_config('lock_timeout', %s, true)", (f"{int(espera * 1000)}ms",))
        cur.execute("SELECT pg_advisory_lock(%s)", (LOCK_MIGRACIONES,))
    conn.commit()


def _soltar_lock(conn: psycopg.Connection) -> None:
    """Suelta el lock. Va en un `finally`, así que puede tocarle una conexión en mal estado.

    Dos motivos para el `rollback` previo: si una migración falló, la transacción está abortada y
    cualquier consulta sobre ella daría `InFailedSqlTransaction`, tapando el error de verdad.
    """
    if conn.info.transaction_status in (TransactionStatus.INERROR, TransactionStatus.UNKNOWN):
        conn.rollback()
    with conn.cursor() as cur:
        cur.execute("SELECT pg_advisory_unlock(%s)", (LOCK_MIGRACIONES,))
    conn.commit()


def _pending(conn: psycopg.Connection, files: list[Path]) -> list[Path]:
    with conn.cursor() as cur:
        cur.execute(_TRACKING_TABLE)
        cur.execute("SELECT version FROM schema_migrations")
        applied = {row[0] for row in cur.fetchall()}
    return [f for f in files if f.name not in applied]


def apply_migrations(
    conn: psycopg.Connection,
    migrations_dir: Path = DEFAULT_MIGRATIONS_DIR,
    lock_wait: float = DEFAULT_LOCK_WAIT,
) -> list[str]:
    """Aplica las migraciones pendientes. Devuelve las versiones aplicadas ahora.

    El lock se toma ANTES de `_pending()`, no alrededor del bucle: el fallo está en leer el
    conjunto de aplicadas y decidir con él, así que envolver solo la escritura no arregla nada.

    Y se suelta aquí dentro, nunca en el que llama: `run.py` sigue usando esta misma conexión para
    la pasada entera, y un lock de sesión sin soltar dejaría al initContainer del web esperando los
    30 minutos que dura una pasada en frío.
    """
    files = sorted(migrations_dir.glob("*.sql"))
    _tomar_lock(conn, lock_wait)
    try:
        pending = _pending(conn, files)
        for f in pending:
            with conn.cursor() as cur:
                cur.execute(f.read_text(encoding="utf-8"))
                cur.execute("INSERT INTO schema_migrations (version) VALUES (%s)", (f.name,))
            conn.commit()
        return [f.name for f in pending]
    finally:
        _soltar_lock(conn)
